### ISSUE TITLE
console - org membership grants default roles

### ISSUE THREAD

**Issue by fvanderbiest** [georchestra/georchestra#1887]
Our current implementation of Roles is incomplete.
Belonging to a given Org should grant users specific Roles. This would ease users management a lot.

**Comment by Vampouille** [comment]
Something like `ROLE_ORG_PSC` if user belong to 'psc' org ?

**Comment by fvanderbiest** [comment]
Why not, but it could be anything more meaningful, related to data, eg `ROLE_WRITE_INSTANCES_LAYER` or `ROLE_READ_BDORTHO` ...

**Comment by Vampouille** [comment]
So we can add roles to an org and those roles will be added to user beloging to that org. 

Maybe we can also automatically add a default role to each role with org name : `ROLE_ORG_PSC`

**Comment by fvanderbiest** [comment]
> Maybe we can also automatically add a default role to each role with org name : ROLE_ORG_PSC

Such automation could be confusing, but to be discussed...

**Comment by fvanderbiest** [comment]
> So we can add roles to an org and those roles will be added to user beloging to that org.

Correct, a user would inherit his org's roles, in addition to his own roles.

**Comment by fvanderbiest** [comment]
> Correct, a user would inherit his org's roles, in addition to his own roles.

This is the way GeoServer user/group + role services works by default, see eg http://docs.geoserver.org/stable/en/user/security/usergrouprole/interaction.html#security-rolesystem-interaction

Unfortunately, we cannot take advantage of it because the authentication mechanism we use in GeoServer uses the HTTP header attribute provided by the SP, see http://docs.geoserver.org/stable/en/user/security/usergrouprole/rolesource.html

**Comment by fvanderbiest** [comment]
To be discussed: how this would interfere with https://github.com/georchestra/georchestra/issues/786#issuecomment-357967099 (namely Orgs having parents...)

**Comment by fvanderbiest** [comment]
In terms of UI, this would imply a new widget on the `/ldapadmin/console/#/users/xxxxx/groups` page showing the list of inherited roles and the corresponding Org, eg:

```
INHERITED ROLES
-------------------------------------------------------------------
|     NAME                                     |         ORG       |
-------------------------------------------------------------------
| WRITE_INSTANCES_LAYER     |         O1         |
| READ_BDORTHO                       |         O2        |
-------------------------------------------------------------------
```

**Comment by fvanderbiest** [comment]
IMO this feature should not be implemented at the security-proxy level (as I first thought), but directly in ldapadmin, fully granting new roles to members of an Org.

This would imply that, as long as a user is member of an Org granting specific roles, he cannot be unlinked with these roles.

**Comment by landryb** [comment]
Fwiw, i second this idea, because we're already doing this since a while, and were planning to reimplement it in our 'new' infra in a similar way.

We have a notion of 'offers' attached to an org, and when an org subscribes to an offer, right now we have some goop (drupal rules) that loops over the org users list and automagically puts those users in the list of roles/rights corresponding to the selected offer.

But this shouldnt prevent an admin to selectively 'remove' roles from an user: ie org 'CD15' subscribed to a full offer, but the data admin in this org wants to be the only one with the role allowing him to order data, or we only want a subset of the org users to have access to cadastrapp..

**Comment by fvanderbiest** [comment]
> But this shouldnt prevent an admin to selectively 'remove' roles from an user

This means roles allocation cannot be dynamic (or it will become very complicated).
This comes with drawbacks : updating the role-list an Org grants will not update roles of users from this Org, because the transfer of roles will be done "one-shot", when user becomes member of Org.

**Comment by landryb** [comment]
> But this shouldnt prevent an admin to selectively 'remove' roles from an user
> 
> This means roles allocation cannot be dynamic (or it will become very complicated).
> This comes with drawbacks : updating the role-list an Org grants will not update roles of users from this Org, because the transfer of roles will be done "one-shot", when user becomes member of Org.

Agreed, there are two orthogonal needs here:
* flexibility of rights for a user (ie grant rights, then selectively remove)
* flexibility of globally handling the rights for all users in a given org at once

The most important thing is to have it properly documented :) I'm not sure we want a knob for this (knobs are bad), maybe more an informational message "you've granted role Z to org X, existing X users foo and bar don't have role Z, do you want to/maybe you should grant them this role too?" (that's of course what we do in drupal now..)

**Comment by fvanderbiest** [comment]
OK, let's sum it up: transfer of roles from Org to members will be done statically, ie "one-shot":
 * when user becomes member of Org or registers as an Org member,
 * when the list of roles an Org grants is modified (which will overwrite any manual role removal at user level)

**Comment by landryb** [comment]
Yes, that seems the most sensible thing to do. Maybe with a warning about the 'overwrite any manual role removal' which could lead to unwanted side effects/surprises...

Imo, it's important to be explicit about it, ie the admin changing org roles must be notified *clearly* of the role changes on the member users.

**Comment by fvanderbiest** [comment]
Had a discussion with Cedric Ansard (Hauts-de-France) today - this feature becomes more and more relevant for his use cases, but he would favor the "dynamic" rather than the "static" scenario.

**Comment by landryb** [comment]
what does "dynamic" vs "static" mean in this context ?

**Comment by fvanderbiest** [comment]
forget my previous comment - what we decided on https://github.com/georchestra/georchestra/issues/1887#issuecomment-404520029 covers the use case.

**Comment by fvanderbiest** [comment]
One more question here.
Say I hold a delegation for roles AAA and BB on org "toto".
Should I be allowed to edit the roles list that org "toto" grants to its users, adding (or removing) only roles AAA and BB ?

**Comment by fvanderbiest** [comment]
We got funding from the Hauts-de-France Region to implement what's been discussed here.
Should we create a GIP to discuss this new feature with the whole community + get a go from the PSC ?

**Comment by fvanderbiest** [comment]
> I'd say yes. It's a major (albeit _a priori_ positive) change. We need to make sure it works with everybody.

Let's go then ...

**Comment by fvanderbiest** [comment]
> Say I hold a delegation for roles AAA and BB on org "toto".
> Should I be allowed to edit the roles list that org "toto" grants to its users, adding (or removing) only roles AAA and BB ?

I'd say "yes".
